source/slang/slang-preprocessor.cpp

```diff
@@ -896,280 +1287,763 @@ static Index _parseMacroArgs(
 
-            sb << nextToken.getContent();
+            // The only interesting addition to the unexpanded case is that we wrap
+            // the stream that "plays back" the argument tokens with a stream that
+            // applies macro expansion to them.
+            //
+            ExpansionInputStream* expansion = new ExpansionInputStream(m_preprocessor, stream);
+            m_currentOpStreams.push(expansion);
         }
+        break;
 
-        // Now re-lex the input
+    case MacroDefinition::Opcode::StringizedParam:
+        {
+            // A macro parameter can also be "stringized" in which case the (unexpanded)
+            // argument tokens will be concatenated and escaped to form the content of
+            // a string literal.
+            //
+            // Much of the initial logic is shared with the other parameter cases above.
+            //
+            Index tokenIndex = op.index0;
+            auto loc = m_macro->tokens.m_tokens[tokenIndex].loc;
 
-        SourceManager* sourceManager = preprocessor->getSourceManager();
+            Index paramIndex = op.index1;
+            SLANG_ASSERT(paramIndex >= 0);
+            SLANG_ASSERT(paramIndex < m_macro->params.getCount());
+            SLANG_ASSERT(paramIndex < m_args.getCount());
 
-        // We create a dummy file to represent the token-paste operation
-        PathInfo pathInfo = PathInfo::makeTokenPaste();
-       
-        SourceFile* sourceFile = sourceManager->createSourceFileWithString(pathInfo, sb.ProduceString());
-        SourceView* sourceView = sourceManager->createSourceView(sourceFile, nullptr, poundPoundToken.getLoc());
+            auto arg = m_args[paramIndex];
+            auto argTokens = m_argTokens.getBuffer();
 
-        Lexer lexer;
-        lexer.initialize(sourceView, GetSink(preprocessor), preprocessor->getNamePool(), sourceManager->getMemoryArena());
+            // We will now iterate over the argument tokens that were passed for
+            // this parameter, and use them to build a string.
+            //
+            auto beginToken = argTokens + arg.beginTokenIndex;
+            auto endToken = argTokens + arg.endTokenIndex;
 
-        SimpleTokenInputStream* inputStream = new SimpleTokenInputStream();
-        initializeInputStream(preprocessor, inputStream);
+            // A stringized parameter is always a `"`-enclosed string literal
+            // (there is no way to stringize things to form a character literal).
+            //
+            StringBuilder builder;
+            builder.appendChar('"');
+            for(auto tokenCursor = beginToken; tokenCursor != endToken; tokenCursor++)
+            {
+                auto token = *tokenCursor;
 
-        inputStream->lexedTokens = lexer.lexAllTokens();
-        inputStream->tokenReader = TokenReader(inputStream->lexedTokens);
+                // Any whitespace between the tokens of argument must be collapsed into
```

Comment:
The long/short of my answer here is "ugh!". Checking in Compiler Explore, what we are doing here with `AfterSpace` doesn't quite work as we need/want it to (at least no matching gcc).

There's basically two ways this could be fixed, neither of which I'm in a hurry to do right away:

First, we could try to hack something up so that the `AfterSpace` flag for the expansion of a macro parameter, stringize, etc. is copied from the original token that started that macro "op" and *not* from the first token of any macro argument it pertains to. This would affect macro parameter substiution (both expanded and unexpanded) as well as stringize and paste operations. The resulting logic could get complicated.

Second, we could preserve the whitespace tokens further through the preprocessor, so that:

* both macro bodies and arguments capture both the semantic tokens and trivia/whitespace as part of the raw argument/body token lists
* when constructing the begin/end ranges for macro arguments or token spans in a macro body we make sure to trim any leading/trailing trivia (but not internal whitespace/trivia)
* when "playing back" a macro, we simply play back all those tokens, both semantically-relevant and trivial
* during stringizing, we iterate over the tokens of the argument including internal whitespace

This latter option seems way cleaner to implement, but funnily enough it actually pushes us further in the direction of needing to support whitespace flowing through the preprocessor.